src/inotify/watcher.cr

```diff
@@ -90,29 +82,50 @@ module Inotify
 
     private def wait_for_event
       spawn do
-        loop { @on_event_callback.call(@event_channel.receive) }
+        loop do
+          event = @event_channel.receive
+          @event_callbacks.each do |proc|
+            proc.call event
+          end
+        end
       end
     end
 
-    private def watch(path : String)
-      wd = LibInotify.add_watch(@fd, path, DEFAULT_WATCH_FLAG)
-      raise "inotify add_watch failed" if wd == -1
-      LOG.debug "inotify add_watch directory #{wd} #{path}"
-      @watch_list[wd] = WatchInfo.new(wd, path, true)
-      unless Dir.empty?(path) || !@recursive
-        Dir.each_child(path) { |child| watch(File.join(path, child)) if File.directory?(File.join(path, child)) }
+    # TODO: block with no event argument ????
+    def on_event(&block : Event ->)
+      @event_callbacks.push block
+    end
+
+    def watch(path : String, mask = DEFAULT_WATCH_FLAG)
+      wd = LibInotify.add_watch(@fd, path, mask)
+      raise Errno.new "inotify add_watch failed" if wd == -1
+      LOG.debug "inotify add_watch #{wd} #{path}"
+      if is_dir = File.directory? path
+        @watch_list[wd] = WatchInfo.new wd, path, is_dir
+        unless Dir.empty?(path) || !@recursive
+          Dir.each_child(path) { |child| watch(File.join(path, child)) if File.directory?(File.join(path, child)) }
+        end
+      else
+        @watch_list[wd] = WatchInfo.new wd, path, false
       end
     end
 
-    private def unwatch
-      @watch_list.each_key do |key|
-        unwatch key
+    def unwatch(path : String)
+      @watch_list.each_value do |info|
+        unwatch info.wd if info.path == path
       end
     end
 
-    private def unwatch(wd : LibC::Int)
-      LibInotify.rm_watch(@fd, wd)
-      @watch_list.delete wd
+    def unwatch(wd : LibC::Int)
+      status = LibInotify.rm_watch(@fd, wd)
+      if status == -1
+        case Errno.value
+        when Errno::EBADF then raise IO::Error.new "fd is not a valid file descriptor"
+        when Errno::EINVAL then raise Errno.new "The watch descriptor wd is not valid; or fd is not an inotify file descriptor"
```

Comment:
It can be:
```cr
case Errno
when .ebadf?  then ...
when .einval? then ...
else          raise ...
end
```

Reply:
This won't work. `Errno` is a class not an enum.

Reply:
you're right